Лабораторна робота №1

Дана послідовність з 0 та 1, яка була отримана зміною кожної цифри
початкової послідовності(яка також була з 0 і 1) з ймовірністю 0.5 для кожної
цифри, незалежно одна від одної. Потрібно знайти оптимальну стратегію для
відшукання суми всіх цифр початкової послідовності. Штрафною функцією
буде різниця між сумами 2 послідовностей в квадраті.

Лабораторну бажано робити на Python.

Відповідно до умови ми маємо K={0,1}^n , X={0,1}^n та D={0, 1, ..., n}

Також ми знаємо, що p(x,k) рівноймовірні, тому вони не беруть участь в формуванні функції помилка

p(x,k)=1/(2^n) * 1/(2^n) = 1/(2^2n)


В своїх реалізації я буду використовувати послідовності довжиною 12(але якщо необхідно можна змінити в змінній seq_len)

In [6]:
import itertools
import numpy as np

seq_len=12
seq=[] # Всі можливі послідовності довжини seq_len

for el in itertools.product([0,1],repeat=seq_len):
    seq.append(el) # генерація всіх можливих послідовностей 
    
print(seq[:6])

[(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0), (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1), (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0), (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1), (0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0), (0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1)]


Далі мені трреба знайти оптимальну стратегію, який в данному випадку нам треба мінімізовувати суму по к :  Σ(p(x,k) * W(k,d))

Через те, що ймовірності однакові, ми фокусуємось на штрафній функції, що згідно умови має бути різниця між сумами 2 послідовностей в квадраті.

Отже, в коді для кожного фіксованого к з К, я обчислюю середню квадратичну різниця між сумами 2 послідовностей.

In [7]:
K=seq
D=seq
loss_sum=[]
for i in range(len(D)):
    loss=[] 
    
    for j in range(len(K)):
        loss_val=((sum(D[i])-sum(K[j]))**2)
        loss.append(loss_val)
#         print(loss)
#     print(f'Sequence:{D[i]},\t sum of sequence:{sum(D[i])},\t sum loss:{sum(loss)}\n')
    loss_sum.append(sum(loss))
        

Через те, що декілька моїх зразків мають значення, що дорівнює мінімуму по показницу середній квадратичній помилці, мені необхідно яка сума послідовності частіше приходить до цього значення. 

In [8]:
min(loss_sum)

12288

In [9]:
min_indx=[]
for i in range(len(loss_sum)):
    if loss_sum[i]==min(loss_sum):
        min_indx.append(i)
        
min_indx[:5]

[63, 95, 111, 119, 123]

In [10]:
np.unique(np.array([sum(D[i]) for i in min_indx]),return_counts=True)

(array([6]), array([924], dtype=int64))

Бачу що лише сума 6 приходить до мінімальної помилки, яка дорівнює 12288, таких комбінацій було 924. 
В підсумку, оптимальне значення для послідовності довжиною 12, є q=6

In [ ]:
вв